pkg/securitycontext/restrictprovider.go

```diff
@@ -0,0 +1,310 @@
+limitations under the License.
+*/
+
+package securitycontext
+
+import (
+	"fmt"
+	"strconv"
+
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/golang/glog"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/fielderrors"
+)
+
+type provider struct {
+	*api.SecurityConstraints
+}
+
+// NewRestrictSecurityContextProvider creates a new security context provider with the default constraints
+func NewRestrictSecurityContextProvider(securityConstraints api.SecurityConstraints) SecurityContextProvider {
+	return &provider{
+		SecurityConstraints: &securityConstraints,
+	}
+}
+
+// ApplySecurityContext ensures that each container in the pod has a security context set and that it
+// complies with the policy defined by the SecurityConstraints of the provider
+func (p *provider) ApplySecurityContext(pod *api.Pod) {
+	for idx := range pod.Spec.Containers {
+		c := &pod.Spec.Containers[idx]
+		glog.V(4).Infof("Applying security constraints to %s", c.Name)
+		p.applySecurityContextToContainer(c)
+	}
+}
+
+// applySecurityContextToContainer applies each section of the security context to the container.  As more options
+// become available they should be added here with corresponding application methods.
+func (p *provider) applySecurityContextToContainer(c *api.Container) {
+	p.applyPrivileged(c)
+	p.applyCapRequests(c)
+	p.applySELinux(c)
+	p.applyRunAsUser(c)
+}
+
+// applyRunAsUser will set the RunAsUser to 0 or p.DefaultSecurityContext.RunAsUser (if DefaultSecurityContext is not nil)
+// if the constraints do not allow run as user requests
+func (p *provider) applyRunAsUser(c *api.Container) {
+	if !p.SecurityConstraints.AllowRunAsUser {
+		if p.DefaultSecurityContext != nil {
+			c.SecurityContext.RunAsUser = p.DefaultSecurityContext.RunAsUser
+		} else {
+			c.SecurityContext.RunAsUser = 0
+		}
+	}
+}
+
+// applySELinux will:
+// 1.  if there are not selinux options on the security constraints: take no action
+// 2.  if there are selinux options on the security constraints AND there are no options defined on the container
+//			AND there is a default security context then use all the default settings
+// 3.  if there are selinux options on the security constraints AND options on the container then check each
+//			setting individually.  If the individual setting is not allowed then remove it or set it to the default if one exists
+func (p *provider) applySELinux(container *api.Container) {
+	// no security context settings for SELinux
+	if p.SecurityConstraints.SELinux == nil {
+		return
+	}
+
+	constraints := p.SecurityConstraints
+	hasDefault := constraints.DefaultSecurityContext != nil
+	hasDefaultSELinux := hasDefault && constraints.DefaultSecurityContext.SELinuxOptions != nil
+
+	// if the container has not defined SELinux options then apply the default if it exists
+	if container.SecurityContext.SELinuxOptions == nil {
```

Comment:
I don't like it because it's an API change - the time for that has come and gone - we will look at it for the next release.

----- Original Message -----
> > +// 2.  if there are selinux options on the security constraints AND there
> > are no options defined on the container
> > +//			AND there is a default security context then use all the default
> > settings
> > +// 3.  if there are selinux options on the security constraints AND
> > options on the container then check each
> > +//			setting individually.  If the individual setting is not allowed then
> > remove it or set it to the default if one exists
> > +func (p *provider) applySELinux(container *api.Container) {
> > +	// no security context settings for SELinux
> > +	if p.SecurityConstraints.SELinux == nil {
> > +		return
> > +	}
> > +
> > +	constraints := p.SecurityConstraints
> > +	hasDefault := constraints.DefaultSecurityContext != nil
> > +	hasDefaultSELinux := hasDefault &&
> > constraints.DefaultSecurityContext.SELinuxOptions != nil
> > +
> > +	// if the container has not defined SELinux options then apply the
> > default if it exists
> > +	if container.SecurityContext.SELinuxOptions == nil {
> 
> I expect @smarterclayton not to like this but i think doing
> `container.SecurityContext` -> `containerContext` and `constraints.SELinux`
> -> `selConstraints` would make this much more readable and you could get
> away without breaking this method up, just a thought.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6287/files#r28998305